lxc/remote.go

```diff
@@ -167,72 +171,107 @@ func (c *cmdRemoteAdd) RunToken(server string, token string, rawToken *api.Certi
 	for _, addr := range rawToken.Addresses {
 		addr = fmt.Sprintf("https://%s", addr)
 
-		conf.Remotes[server] = config.Remote{Addr: addr, Protocol: c.flagProtocol, AuthType: c.flagAuthType, Domain: c.flagDomain}
-
-		_, err = conf.GetInstanceServer(server)
+		err := c.addRemoteFromToken(addr, server, token, rawToken.Fingerprint)
 		if err != nil {
-			certificate, err = shared.GetRemoteCertificate(addr, c.global.conf.UserAgent)
-			if err != nil {
+			if errors.Is(err, &serverAddressUnavailableError{}) {
 				continue
 			}
 
-			certDigest := shared.CertFingerprint(certificate)
-			if rawToken.Fingerprint != certDigest {
-				return fmt.Errorf(i18n.G("Certificate fingerprint mismatch between certificate token and server %q"), addr)
-			}
+			return err
+		}
 
-			dnam := conf.ConfigPath("servercerts")
-			err := os.MkdirAll(dnam, 0750)
-			if err != nil {
-				return fmt.Errorf(i18n.G("Could not create server cert dir"))
-			}
+		return nil
+	}
 
-			certf := conf.ServerCertPath(server)
+	fmt.Println(i18n.G("All server addresses are unavailable"))
+	fmt.Printf(i18n.G("Please provide an alternate server address (empty to abort):") + " ")
 
-			certOut, err := os.Create(certf)
-			if err != nil {
-				return fmt.Errorf(i18n.G("Failed to create %q: %w"), certf, err)
-			}
+	line, err := shared.ReadStdin()
+	if err != nil {
+		return err
+	}
 
-			err = pem.Encode(certOut, &pem.Block{Type: "CERTIFICATE", Bytes: certificate.Raw})
-			if err != nil {
-				return fmt.Errorf(i18n.G("Failed to write server cert file %q: %w"), certf, err)
-			}
+	if len(line) == 0 {
+		return fmt.Errorf(i18n.G("Failed to add remote"))
+	}
 
-			err = certOut.Close()
-			if err != nil {
-				return fmt.Errorf(i18n.G("Failed to close server cert file %q: %w"), certf, err)
-			}
+	err = c.addRemoteFromToken(string(line), server, token, rawToken.Fingerprint)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (c *cmdRemoteAdd) addRemoteFromToken(addr string, server string, token string, fingerprint string) error {
+	conf := c.global.conf
+
+	var certificate *x509.Certificate
+	var err error
+
+	conf.Remotes[server] = config.Remote{Addr: addr, Protocol: c.flagProtocol, AuthType: c.flagAuthType, Domain: c.flagDomain}
+
+	_, err = conf.GetInstanceServer(server)
+	if err != nil {
+		certificate, err = shared.GetRemoteCertificate(addr, c.global.conf.UserAgent)
+		if err != nil {
+			return &serverAddressUnavailableError{}
```

Comment:
Using `api.StatusErrorf(http.StatusServiceUnavailable, "%v", err)` would avoid the need for a custom error type, and you can match it using `api.StatusErrorCheck(err, http.StatusServiceUnavailable)`.